### PR TITLE
Fix coupon usage limits on failed trashed orders

### LINKED ISSUE #31245: Coupon usage limits and failed orders.
**Prerequisites (mark completed items with an [x]):**
- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
A user has a coupon code that has either the `Usage limit per coupon` or `Usage limit per user` set to `1`.
If they try to checkout with the coupon and the payment fails, the coupon gets stuck. The next retry fails and removes the coupon, even if payment would have succeeded. The coupon is no longer usable.
This issue is reproducible on both the woocommerce.com platform and on a fresh installation.

It looks like the `WC_Discounts::validate_coupon_usage_limit()` and `WC_Discounts::validate_coupon_user_usage_limit` methods don't account for re-trying an order that already has a coupon attached.

**Expected behavior**
The coupon doesn't get stuck on payment retries. It was already successfully applied to the order being checked out. 

**Actual behavior**
The coupon gets stuck on the first payment failure and get's removed on retry. The coupon is no longer usable.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
- Add a coupon with `Usage limit per coupon` or `Usage limit per user` set to `1`.
- Add a product to the cart and proceed to checkout.
- In the cart or checkout pages add the coupon.
- Try to complete the checkout with a Stripe test card that fails payment, like `4000000000000002` which returns a `card_declined` error. The payment will fail.
- Now try using a test card that works like `4242424242424242`.
- See an error at the top claiming there's a problem with the coupon and confirm that the coupon was removed.
- Trying to re-add the coupon fails with the same error.
- The coupon is now stuck and can't be used even though `wp-admin` reports `0/1` usages.
- Canceling the failed order releases the coupon.

**Screenshots**

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**

### WordPress Environment ###

WordPress address (URL): http://localhost:10018
Site address (URL): http://localhost:10018
WC Version: 5.9.0
REST API Version: ✔ 5.9.0
WC Blocks Version: ✔ 6.1.0
Action Scheduler Version: ✔ 3.3.0
WC Admin Version: ✔ 2.8.0
Log Directory Writable: ✔
WP Version: 5.8.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 7.3.5
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.54.0
LibreSSL/2.6.5

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.6.0
WC Database Prefix: wp_
Total Database Size: 4.41MB
Database Data Size: 2.96MB
Database Index Size: 1.45MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.05MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_deposits_payment_plans: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_deposits_payment_plans_schedule: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 2
product: 1
shop_coupon: 1
shop_order: 3

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (3) ###

WooCommerce Deposits: by WooCommerce – 1.5.6
WooCommerce Stripe Gateway: by WooCommerce – 5.7.0
WooCommerce: by Automattic – 5.9.0

### Inactive Plugins (2) ###

WooCommerce Payments: by Automattic – 2.6.0 (update to version 3.2.3 is available)
WooCommerce PayPal Payments: by WooCommerce – 1.3.2 (update to version 1.6.1 is available)

### Must Use Plugins (1) ###

local.php: by  –

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-One
Version: 1.3 (update to version 1.4 is available)
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 18
Oldest: 2021-11-18 11:44:16 +0000
Newest: 2021-11-18 12:39:36 +0000

Pending: 1
Oldest: 2021-11-19 11:45:27 +0000
Newest: 2021-11-19 11:45:27 +0000

### Status report information ###

Generated at: 2021-11-18 12:56:33 +00:00
`

### PR DESCRIPTION
**Pull request by sflems**
### Submission Review Guidelines:

- [x] I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- [x] I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

This addresses coupon usage limits for failed & trashed orders. Specifically this issue below is prevalent when the block checkout creates a pending order, which then fails. The coupon usage is not updated as expected. It may rectify other similar cases as well.

- The changes to `wc_update_coupon_usage_counts` include adding actions  for`woocommerce_order_status_failed` & `woocommerce_trash_order`.
- A new filter `woocommerce_update_coupon_usage_invalid_statuses` has been added, which now includes default `cancelled`, `trash`, and `failed` statuses. This allows filtering should actions be added for custom order statuses. 

Closes #31245.
Possibly related to #33648.

#### wp-env

1. `pnpm --filter=@woocommerce/plugin-woocommerce test:unit:env -- --filter={WC_Order_Functions_Test,WC_Abstract_Order_Test}`

#### Manually

1. Ensure your test site has the "Cash on delivery" payment method enabled, as this is the easiest one to test with.
1. Create a coupon.
2. Add a product to your cart and then apply the coupon.
3. Add the following snippet to your test site to simulate a failed COD order:
	```php
	function test_wc_cod_order_status( $status ) {
		return 'failed';
	}
	add_filter( 'woocommerce_cod_process_payment_order_status', 'test_wc_cod_order_status', 15 );
	```
5. Go through the checkout process and use the Cash on delivery payment method. You should get a message saying the payment failed.
6. On another tab, visit the Coupons screen in WP Admin. The coupon should still show `0` uses. (Without the change in this PR, it would now show `1`).
7. Disable/remove the code snippet added above.
8. Back on the checkout screen, click "Try again" and complete the payment. With the code snippet removed, this should now succeed.
9. On the Coupons screen, the coupon should now show `1` use.
10. Go to the Orders screen, and trash the order that you just completed.
11. Back on the Coupons screen, the coupon should now once again show `0` uses.

### Other information:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes, as applicable?
- [x] Have you created a changelog file for each project being changed, ie pnpm --filter=<project> changelog add?
- [x] Have you included testing instructions?

#### Significance

-   [x] Minor

#### Type

-   [x] Fix - Fixes an existing bug

#### Message 

Update coupon_usage for failed & trashed orders.

### CODE REVIEW

File: plugins/woocommerce/includes/wc-order-functions.php [hunk 1]
```diff
@@ -934,14 +934,27 @@ function wc_update_coupon_usage_counts( $order_id ) {
 	}
 
 	$has_recorded = $order->get_data_store()->get_recorded_coupon_usage_counts( $order );
+	$invalid_statuses = array( 'cancelled', 'failed', 'trash' );
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/wc-order-functions.php]:
```suggestion
	$has_recorded     = $order->get_data_store()->get_recorded_coupon_usage_counts( $order );
	$invalid_statuses = array( 'cancelled', 'failed', 'trash' );
```

Appease the code style linter

File: plugins/woocommerce/tests/php/includes/wc-order-functions-test.php [hunk 2]
```diff
@@ -110,4 +110,70 @@ public function test_wc_update_total_sales_counts() {
...
 
+
+	/**
+	 * Test wc_update_coupon_usage_counts and check usage_count after order reflection.
+	 *
+	 * Tests the fix for issue #31245
+	 */
+	public function test_wc_update_coupon_usage_counts() {
+
+		$coupon = WC_Helper_Coupon::create_coupon( 'test' );
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/includes/wc-order-functions-test.php:121]:
```suggestion
	public function test_wc_update_coupon_usage_counts() {
		$coupon = WC_Helper_Coupon::create_coupon( 'test' );
```

Appease the code style linter.

File: plugins/woocommerce/tests/php/includes/wc-order-functions-test.php [hunk 3]
```diff
@@ -110,4 +110,70 @@ public function test_wc_update_total_sales_counts() {
...
+		$order_id = WC_Checkout::instance()->create_order(
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/includes/wc-order-functions-test.php]:
```suggestion
	public function test_wc_update_coupon_usage_counts() {
		$coupon   = WC_Helper_Coupon::create_coupon( 'test' );
		$order_id = WC_Checkout::instance()->create_order(
```

Appease the code style linter.

**Review by coreymckrill — APPROVED** [review]
Thanks for the PR @sflems! This tested well on my local dev setup. And thanks especially for writing good testing instructions and including unit tests. ⭐ 

There are a couple of tweaks necessary to make the code style linter happy that I'll just go ahead and commit. I also updated the testing instructions to be a little bit more explicit to help the folks that do additional testing of everything before a release.

**Comment by coreymckrill** [comment]
I'm going to do a bit more testing using the reproduction steps in #31245 (i.e. with Stripe) before merging...

**Comment by coreymckrill** [comment]
Confirmed that with this changeset, following the reproduction steps in #31245 using Stripe payment method, the failed payment no longer uses up the coupon limit, and the coupon is no longer stuck when trying again with a different card!

**Comment by sflems** [comment]
Right on, this bug looks to have reared its head here and there in the past, so great to have it under wraps. Thank you for the review!

**Pull request merged**